kernel/src/fs/notify/inotify.rs

```diff
@@ -0,0 +1,607 @@
+            if let (Some(inode), Some(subscriber)) =
+                (entry.inode.upgrade(), entry.subscriber.upgrade())
+            {
+                // Remove the subscriber from the inode's publisher.
+                inode.fsnotify_publisher().remove_subscriber(&subscriber);
+                // Remove the subscriber from the file system's fsnotify info.
+                inode.fs().fsnotify_info().remove_subscriber();
+            }
+        }
+    }
+}
+
+/// Default max queued events.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.14/source/fs/notify/inotify/inotify_user.c#L853>
+const DEFAULT_MAX_QUEUED_EVENTS: usize = 16384;
+const INOTIFY_EVENT_HEADER_SIZE: usize = core::mem::size_of::<u32>() * 4;
+
+impl InotifyFile {
+    /// Creates a new inotify file.
+    pub fn new(is_nonblocking: bool) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            next_wd: AtomicU32::new(1),
+            watch_map: RwLock::new(HashMap::new()),
+            watch_lock: Mutex::new(()),
+            is_nonblocking: AtomicBool::new(is_nonblocking),
+            event_queue: RwLock::new(VecDeque::new()),
+            queue_capacity: DEFAULT_MAX_QUEUED_EVENTS,
+            this: weak_self.clone(),
+            pollee: Pollee::new(),
+        })
+    }
+
+    /// Allocates a new watch descriptor.
+    fn alloc_wd(&self) -> Result<u32> {
+        const MAX_VALID_WD: u32 = i32::MAX as u32;
+
+        let new_wd = self.next_wd.fetch_add(1, Ordering::Relaxed);
+        if new_wd > MAX_VALID_WD {
+            // Rollback the allocation if we exceeded the limit
+            self.next_wd.fetch_sub(1, Ordering::Relaxed);
+            return_errno_with_message!(Errno::ENOSPC, "Inotify watches limit reached");
+        }
+        Ok(new_wd)
+    }
+
+    /// Adds or updates a watch on a path.
+    ///
+    /// If a watch on the path is not found, creates a new watch.
+    /// If a watch on the path is found, updates it.
+    pub fn add_watch(
+        &self,
+        path: &Path,
+        interesting: InotifyEvents,
+        options: InotifyControls,
+    ) -> Result<u32> {
+        // Serialize updates so concurrent callers do not create duplicate watches.
+        let _guard = self.watch_lock.lock();
+
+        // Try to update existing subscriber first
+        match self.update_existing_subscriber(path, interesting, options) {
+            Ok(wd) => Ok(wd),
+            Err(e) if e.error() == Errno::ENOENT => {
+                // Subscriber not found, create a new one
+                self.create_new_subscriber(path, interesting, options)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Removes a watch by watch descriptor.
+    pub fn remove_watch(&self, wd: u32) -> Result<()> {
+        let _guard = self.watch_lock.lock();
+
+        // Find the subscriber entry by the watch descriptor.
+        let Some(entry) = self.watch_map.read().get(&wd).cloned() else {
```

Comment:
So why do we want to do `watch_map.read().get()` in the beginning of this method and then do `self.watch_map.write().remove(&wd)` in the end... we should have called `remove` in the first place to obtain the removed `entry`.